### ISSUE TITLE
Create an initial "Collection" resource type

### ISSUE THREAD

**Issue by danames** [hydroshare/hydroshare#839]
We (BYU) are going to create an initial version of a "collection" resource type. It will include the following steps:

- [ ] Use the web app resource type as a template since it doesn't require uploading a file and neither will collection.
- [ ] The add a "resource specific meta data element" that is a list of resources (by GUID) that are grouped using this collection
- [ ] To create the resource, the user will need to click the edit button on the resource landing page and go to the resource specific metadata tab and see a list of all resources that are in their "My Resources" list. 
- [ ] User can click the check box next to some of their resources in that list
- [ ] The GUID's for the clicked resources get saved in the metadat file and hence also the Bagit file for the collection resource.
- [ ] Once this much works, then we will create a new issue and design for a web app in Tethys that allows a user to aggregate their resources on a map or in other visual ways and create a collection resource from it. 
- [ ]  Much later we could build a function into the My Resources landing page such that the user just clicks some resources and clicks a "Create Collection" Button.

**Comment by horsburgh** [comment]
Seems like the much easier way to go about this for users would be to add a button to the My Resources page. Then, whichever resources are selected on the My Resources page would be added to a new resource that is a collection. Why not do that first?

Also - the list of resources aggregated into the collection should be set in the resource map document and not in the resourcemetadata.xml file. 

Another thought - since the list of resources that are part of the collection is essentially the content of the resource, it might seem odd to put that list in the resource specific metadata section on the landing page. I think it should just occupy the content section of the resource landing page in the same way regular content files do.

**Comment by danames** [comment]
We've discussed this first suggestion, and actually it is part of the list
of to do items on the issue (the last entry). However, it seems like it is
better to get the resource type built correctly first, and then work on
ways to build the resource. We have identified 3 ways: 1) user edits the
resource on the resource landing page. 2) user builds a map with various
data layers and other linked resources in Tethys and saves it to a new
collection resource. 3) user selects the resources on the my resources page
and saves them to a new resource.

I agree we could save the list in the resource map instead of the resource
metadata.

Can we edit/replace the content section of the resource landing page? If so
then I agree that's where the list of linked resources should go.

**Comment by horsburgh** [comment]
I can talk to @mauriel133 about what we can do with the content section. It will be important to figure this out for both collections and "composite" resources (if we eventually want to support those).

**Comment by danames** [comment]
What is the difference between a collection and a composite?

**Comment by horsburgh** [comment]
Composite resources may have content files but also reference other resources.

**Comment by selimnairb** [comment]
@danames We should use foreign keys to refer to resources from within the HydroShare database, not GUIDS.

**Comment by horsburgh** [comment]
Not sure about all of the details, but the resource map would use the full URL to the resources that are aggregated into the collection.

**Comment by ironfroggy** [comment]
I want to make both an implementation recommendation and also raise concern about security and permission issues as they apply to this ticket. I'm making my implementation recommendation with the security and permission issues in mind, so hopefully if you like what I suggest we can address that concern.

First, in regards to the concern, we have to think about how the permissions of the collection and the permissions of the resources *in* the collection interact. There are both obvious and non-obvious cases.

- A user creating a collection needs permissions on a resource to add it to the collection. At the very least they need permission to view it.
- Does a user need permission to edit a resource to add it to a collection?
- A user viewing a collection needs permission to view each resource in that collection in order to see the contents of the collection.
- If a user does not have view permission on any of the resources in a collection, should they see the collection at all?
- If a user has view permission on only *some* of the resources in the collection, how should that be presented? Should the collection simply appear smaller or should they be told there are resources hidden from them due to permissions?
- Is it reasonable to require a user needs view permissions for all resources in a collection to view the collection at all, so everyone sees the same collection contents?
- Does granting view permissions to a collection grant view permission to all its contents?
- Does removing view permissions to a resource for a user remove that resource from collections they have put it in?

Secondly, I'd like to recommend some guidance on implementation.

I'd agree with @selimnairb on references over GUIDs, it will perform much better and make it very easy to combine a query against a collections' ManyToMany column with a query from Alva's permission system, so that we can ensure a proper security model with regard to the permissions of users inside collections. We need to answer some of the questions raised above, and depending on what the answers are we might want guidance from him on expanding those routines to cover some of the specific cases (like if we want to require permissions on *all* contents of a collection).

In particular I'd advise against denormalization of the collection, because doing so would make it very easy to subvert the permissions system by mistake.

**Comment by danames** [comment]
>
>
>    - A user creating a collection needs permissions on a resource to add
>    it to the collection. At the very least they need permission to view it.
>
> Correct

>
>    - Does a user need permission to edit a resource to add it to a
>    collection?
>
> No. As long as they have "view" permission, they can add it to a
collection.

>
>    - A user viewing a collection needs permission to view each resource
>    in that collection in order to see the contents of the collection.
>
> Yes.

>
>    - If a user does not have view permission on any of the resources in a
>    collection, should they see the collection at all?
>
> Yes they should see the collection.

>
>    - If a user has view permission on only *some* of the resources in the
>    collection, how should that be presented? Should the collection simply
>    appear smaller or should they be told there are resources hidden from them
>    due to permissions?
>
> They should be told there are some resources in the collection that they
can't view because of permissions.

>
>    - Is it reasonable to require a user needs view permissions for all
>    resources in a collection to view the collection at all, so everyone sees
>    the same collection contents?
>
> No. See above.

>
>    - Does granting view permissions to a collection grant view permission
>    to all its contents?
>
> View permissions on a collection are based on view permissions of the
contents.

>
>    - Does removing view permissions to a resource for a user remove that
>    resource from collections they have put it in?
>
> No, it just makes the resource invisible to users who don't have view
permissions to it.

>
>
> Secondly, I'd like to recommend some guidance on implementation.
>
> I'd agree with @selimnairb <https://github.com/selimnairb> on references
> over GUIDs, it will perform much better and make it very easy to combine a
> query against a collections' ManyToMany column with a query from Alva's
> permission system, so that we can ensure a proper security model with
> regard to the permissions of users inside collections. We need to answer
> some of the questions raised above, and depending on what the answers are
> we might want guidance from him on expanding those routines to cover some
> of the specific cases (like if we want to require permissions on *all*
> contents of a collection).
>
> In particular I'd advise against denormalization of the collection,
> because doing so would make it very easy to subvert the permissions system
> by mistake.
>

Thanks for thinking through these issues. Sarva Pulla in my group is
working on a prototype of the collection and this is useful informatino for
him going forward.

**Comment by ironfroggy** [comment]
If removing a user's permission to view a resource does not remove that resource from collections they have put it in, how would that resource be removed from the collection? The user who owns the collection would not have permission to see the resource in order to remove it.

**Comment by danames** [comment]
Perhaps we always show the names of the resources in the list of resources
in a collection. If the person viewing or building the collection doesn't
have permission to view one of the collected resources, then he/she just
sees the name. If he/she is the creator of this collection he/she can
remove it by its name.

Does this violate any rules we've already created? I.e. do we have a rule
to not show resource names to people who don't have permission to view the
resource files?

**Comment by hyi** [comment]
Just a quick comment on showing resource names. By resource names, I assume they mean resource titles. Since resource titles are not unique meaning multiple resources could end up with the same titles, revealing resource titles/names alone are not sufficient to distinguish resources.

**Comment by ironfroggy** [comment]
Do any of these plans account for the need for other types of resources that aggregate other resources? Collections are kind of the generic form of this, but I understand there will also be some sorts of "composite" resource types. It would be really helpful for that stage to consider how implementation of collections can be reused for these composite types.

**Comment by horsburgh** [comment]
As @ironfroggy states - it might be useful to think through "collection" versus "composite".  They might potentially be the same thing (or at least maybe one implementation could support both).

**Comment by danames** [comment]
Here are a few use cases to help frame the discussion:

Use case 1: Group together different map/shape files associated with a map
for display in ArcMap. Download the group of files all at once directly
into ArcMap

Use case 2: Group together time series, maps, and models associated with a
specific engineering project. Share this group of resources with
stakeholders interested in the project. Stakeholders can comment on the
collection and discuss in the comments.

Use case 3: Group together the data files and reports associated with an
NSF funded research project. Publish the group/collection such to meet NSF
requirements. Assign a new DOI to the collected material.

**Comment by dtarb** [comment]
Collections/composite resources are described in the resource data model document (see https://docs.google.com/document/d/1_XY7J8hXQl-w24FWyuWs7YfRHw68px1ZOB_dQOU7sCQ/edit).  I have just migrated this to Google to make this (a bit) easier than locating in Alfresco.

We need to address whether the use cases demand inclusion by reference only, or inclusion by value only or both.  Inclusion by value may support a publication/reproducibility use case where you want to preserve the actual copy of the data you used notwithstanding changes to the original resource.

**Comment by zhiyuli** [comment]
@selimnairb @ironfroggy
About storing the GUIDs or using the Foreign Key, our quick test showed the we may use ManyToManyField in the models design like:
collection = models.ManyToManyField(BaseResource, null=True, blank=True)
A collection can have multiple member resources; A certain resource can be members of multiple collections. 
One frontend, short_id (GUID) is the only identifier for a certain resource. So the user action 'selecting some resources to create a collection' is actually selecting some resource short_ids and sending these short_ids to backend. Backend retrieves resource objects by res=BaseResource.objects.filter(short_id__iexact=res_id). Then, use collection.add(res) to establish the ManyToMany relationship. Is this the right way to do it? Thanks

@danames  As a collection itself is a resource, can a collection be a member of another collection? Thanks

**Comment by danames** [comment]
If it's simple to implement, then there is no reason not to allow a
collection to hold other collections.

**Comment by zhiyuli** [comment]
@danames  @SarvaPulla 
If we allow a collection to hold other collections.A new situation should be handled: 
If collection A is holding collection B, we should not allow collection B to hold collection A.
Further, we dont allow collection A to hold collection B if collection B itself, or its any collection-type member, or member's member already holds collection A.

**Comment by danames** [comment]
OK good point. I think for starters we should not allow collections to hold
collections. This will save a lot of potential complications. If some time
down the road we need this capability then we can add it then.

**Comment by zhiyuli** [comment]
@danames @dtarb @horsburgh @alvacouch @SarvaPulla 

We just deployed the latest collction code on playground.  Link to a public collection resource:  https://playground.hydroshare.org/resource/7eae875268a14c9c9b0edbfad0d25f8c/

Some notes:
1) We are still using the "complex" permission rules we demo'd two weeks ago;

2) The interface has been changed based on @horsburgh suggestions. The collection contents list on landing page directly, not on resource-specific tab page. The resources that can be added into this collection are shown in a popup window. 
Styling work is not done yet, especially the "Add" button and the popup window. (need help from @mauriel133 ).

3) We disable the "Publish" button on collection landing page as we have not decided what can be published yet. 

thanks

**Comment by danames** [comment]
Hi all, just a quick note about collections. Please test it on playground
and give feedback soon. We need this to be deployed to www sooner than
later for some real use cases and to support something new and exciting we
are going to show you at the AHM. - Dan

**Comment by dtarb** [comment]
I tried to write a comment on the example collection @zhiyuli provided and got the error below.  As a result I am still skeptical (based on track record and this test) of the readiness to rush this to www deploy sooner rather than later.  For something new and exciting (and mysterious and untested) for the AHM @danames I would suggest we all agree to use beta.hydroshare.org for these things so that we are experimenting on ourselves, not on the operational www system.  Michael, could you arrange for beta to have the latest www resources and then could the team agree to do demo's of new things on beta.  It does not have to be beta - it could be any of our test environments, as long as we all agree.  But I do not think we should be rushing untested things to www, for the purposes of AHM demos.  Sorry @danames but when you have established a reputation for deploys that I can not break in less than 2 min, then I may soften this stance.

IntegrityError at /comment/

duplicate key value violates unique constraint "django_comments_pkey"
DETAIL:  Key (id)=(12) already exists.

Request Method: 	POST
Request URL: 	https://playground.hydroshare.org/comment/
Django Version: 	1.7.8
Exception Type: 	IntegrityError
Exception Value: 	

duplicate key value violates unique constraint "django_comments_pkey"
DETAIL:  Key (id)=(12) already exists.

Exception Location: 	/usr/local/lib/python2.7/dist-packages/django/db/backends/utils.py in execute, line 65
Python Executable: 	/usr/local/bin/uwsgi
Python Version: 	2.7.6

**Comment by zhiyuli** [comment]
@dtarb 
Sorry you got an error on the comment feature, but I dont think this error is related to collection. I just tried to leave a comment for a generic resource on playground, it also failed. Anyway you could paste your comment about the collection here and we will continue improving it. 

For the broken comment feature, i would like to defer to @hyi or @pkdash as they know more about it than I do.

**Comment by dtarb** [comment]
Consider whether "Authors" should be listed as part of the collection.  Authors are the intellectual creators of the thing being listed, so are arguably more important than Owners.  In many cases they will be the same.

Also, when some of the resources are geographic consider having a coverage element that is the union of the coverage elements of the geographic resources.

**Comment by hyi** [comment]
@dtarb @zhiyuli is right, this is not a problem specific to collection implementation. I tried on playground to post a comment to https://playground.hydroshare.org/resource/1302db3c1a76475ea7e87d7ba881f549 and got the integrity error initially, then I tried to click New comment button, and the comment got posted successfully without issues. It looks like every time you post a comment, the primary key gets increased by 1, if the key already exists which raises integrity error (which should not happen in the first place), and you tried to figure out what was going on and tried to debug it, then the problem went away since primary key increases by 1 each time and usually before you could figure out the issue, the issue never happen any more. Since we are pretty much using Mezzanine's commenting feature out of the box, it is not straightforward to figure out the issue and requires digging into Mezzanine's codebase, but after one or two tries, the issue went away and never occur any more. Due to its "unreproducible" nature and Mezzanine's "hidden" code, I have never found the time to dig into it. Perhaps it is a good idea to create an issue about it just to keep track of it.

**Comment by dtarb** [comment]
OK.  However I still think we should not rush untested and unseen features to www for demo purposes.  We should use beta or another testing platform.

**Comment by danames** [comment]
If anyone has a legitimate reason not to graduate "collections" to www,
please let me know. I think the code is as robust and ready to go as any of
the other resource types - though I'm sure there will be opportunities to
improve it, as with other resource types.

The demo we have planned is a new experimental app that displays/views
multiple resources together - optimally read from a collection resource. We
can run this app against dev or playground, but, again, I don't see any
reason not to move collections to www sooner than later.

**Comment by pkdash** [comment]
There is a PR (https://github.com/hydroshare/hydroshare/pull/1000) for this issue which is yet to be reviewed due to its WIP status. Once it is merged to develop, it would be good to stage it for testing before moving it to www.

**Comment by zhiyuli** [comment]
closed. the init version (phase 1) was already merged
